crates/core/src/error.rs

```diff
@@ -1,810 +1,451 @@
+//!
+//! ### Traits
+//!
+//! All traits and implementations of (foreign) traits should use `RusticResult` as return type or `Box<RusticError>` as `Self::Err`.
+//!
+//! ### Display and Debug
+//!
+//! All types that we want to attach to an error should implement `Display` and `Debug` to provide a good error message and a nice way
+//! to display the error.
 
 // FIXME: Remove when 'displaydoc' has fixed/recommended further treatment upstream: https://github.com/yaahc/displaydoc/issues/48
 #![allow(clippy::doc_markdown)]
 // use std::error::Error as StdError;
 // use std::fmt;
 
+use ::std::convert::Into;
+use smol_str::SmolStr;
 use std::{
-    error::Error,
-    ffi::OsString,
-    num::{ParseFloatError, ParseIntError, TryFromIntError},
-    ops::RangeInclusive,
-    path::{PathBuf, StripPrefixError},
-    process::ExitStatus,
-    str::Utf8Error,
+    backtrace::Backtrace,
+    borrow::Cow,
+    fmt::{self, Display},
 };
 
-#[cfg(not(windows))]
-use nix::errno::Errno;
+pub(crate) mod constants {
+    pub const DEFAULT_DOCS_URL: &str = "https://rustic.cli.rs/docs/errors/";
+    pub const DEFAULT_ISSUE_URL: &str = "https://github.com/rustic-rs/rustic_core/issues/new";
+}
 
-use aes256ctr_poly1305aes::aead;
-use chrono::OutOfRangeError;
-use displaydoc::Display;
-use thiserror::Error;
+/// Result type that is being returned from methods that can fail and thus have [`RusticError`]s.
+pub type RusticResult<T, E = Box<RusticError>> = Result<T, E>;
 
-use crate::{
-    backend::node::NodeType,
-    blob::{tree::TreeId, BlobId},
-    repofile::{indexfile::IndexPack, packfile::PackId},
-};
+/// Severity of an error, ranging from informational to fatal.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Severity {
+    /// Informational
+    Info,
 
-/// Result type that is being returned from methods that can fail and thus have [`RusticError`]s.
-pub type RusticResult<T> = Result<T, RusticError>;
+    /// Warning
+    Warning,
 
-// [`Error`] is public, but opaque and easy to keep compatible.
-#[derive(Error, Debug)]
-#[error(transparent)]
-/// Errors that can result from rustic.
-pub struct RusticError(#[from] pub(crate) RusticErrorKind);
+    /// Error
+    Error,
 
-// Accessors for anything we do want to expose publicly.
-impl RusticError {
-    /// Expose the inner error kind.
-    ///
-    /// This is useful for matching on the error kind.
-    pub fn into_inner(self) -> RusticErrorKind {
-        self.0
-    }
+    /// Fatal
+    Fatal,
+}
 
-    /// Checks if the error is due to an incorrect password
-    pub fn is_incorrect_password(&self) -> bool {
-        matches!(
-            self.0,
-            RusticErrorKind::Repository(RepositoryErrorKind::IncorrectPassword)
-        )
-    }
+/// Status of an error, indicating whether it is permanent, temporary, or persistent.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Status {
```

Comment:
just a note: not yet used

Reply:
Yes, same as severity. I was unsure where to use it. We can also remove it. If we think we may never use it.

Reply:
That being said, we might want others also to construct errors, e.g. when they implement some backend related things. So we may need different status and severity information there. Maybe not the one, I took for now. E.g. informational might not be useful for errors. But maybe useful for `RusticIssues`, the ones we want to deal with for `CheckCommand`.